daemon/api_validate.go

```diff
@@ -0,0 +1,260 @@
+	if sequence != 0 {
+		key = fmt.Sprintf("%s=%d", key, sequence)
+	}
+	return key
+}
+
+func validationSetNotFound(accountID, name string, sequence int) Response {
+	res := &errorResult{
+		Message: "validation set not found",
+		Kind:    client.ErrorKindValidationSetNotFound,
+		Value:   validationSetKeyString(accountID, name, sequence),
+	}
+	return &resp{
+		Type:   ResponseTypeError,
+		Result: res,
+		Status: 404,
+	}
+}
+
+func validationSetAccountAndName(validationSet string) (accountID, name string, err error) {
+	parts := strings.Split(validationSet, "/")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("invalid validation-set argument")
+	}
+	accountID = parts[0]
+	name = parts[1]
+	if !asserts.IsValidAccountID(accountID) {
+		return "", "", fmt.Errorf("invalid account ID %q", accountID)
+	}
+	if !asserts.IsValidValidationSetName(name) {
+		return "", "", fmt.Errorf("invalid validation set name %q", name)
+	}
+	return accountID, name, nil
+}
+
+func listValidationSets(c *Command, r *http.Request, _ *auth.UserState) Response {
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	validationSets, err := assertstate.ValidationSets(st)
+	if err != nil {
+		return InternalError("accessing validation sets failed: %v", err)
+	}
+
+	names := make([]string, 0, len(validationSets))
+	for k := range validationSets {
+		names = append(names, k)
+	}
+	sort.Strings(names)
+
+	results := make([]validationSetResult, len(names))
+	for i, vs := range names {
+		tr := validationSets[vs]
+		// TODO: evaluate against installed snaps
+		var valid bool
+		modeStr, err := modeString(tr.Mode)
+		if err != nil {
+			return InternalError(err.Error())
+		}
+		results[i] = validationSetResult{
+			ValidationSet: validationSetKeyString(tr.AccountID, tr.Name, tr.PinnedAt),
+			Mode:          modeStr,
+			Seq:           tr.Current,
+			Valid:         valid,
+		}
+	}
+
+	return SyncResponse(results, nil)
+}
+
+func getValidationSet(c *Command, r *http.Request, _ *auth.UserState) Response {
+	vars := muxVars(r)
+	accountID := vars["account"]
+	name := vars["name"]
+
+	if !asserts.IsValidAccountID(accountID) {
+		return BadRequest("invalid account ID %q", accountID)
+	}
+	if !asserts.IsValidValidationSetName(name) {
+		return BadRequest("invalid name %q", name)
+	}
+
+	query := r.URL.Query()
+
+	// sequence is optional
+	sequenceStr := query.Get("sequence")
+	var sequence int
+	if sequenceStr != "" {
+		var err error
+		sequence, err = strconv.Atoi(sequenceStr)
+		if err != nil {
+			return BadRequest("invalid sequence argument")
+		}
+		if sequence < 0 {
+			return BadRequest("invalid sequence argument: %d", sequence)
+		}
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var tr assertstate.ValidationSetTracking
+	err := assertstate.GetValidationSet(st, accountID, name, &tr)
+	if err == state.ErrNoState || (err == nil && sequence != 0 && sequence != tr.PinnedAt) {
+		return validationSetNotFound(accountID, name, sequence)
```

Comment:
if we don't find it locally we need to try to look for it in the store, that can be a TODO for now